### PR TITLE
CBG-635: Fix TestHandleProfiling and TestHandleHeapProfiling te…

### PR DESCRIPTION
**Pull request by sarathkumarsivan**

### CODE REVIEW

File: rest/server_context.go [hunk 1]
```diff
@@ -40,12 +41,13 @@ const DefaultStatsLogFrequencySecs = 60
...
-	statsContext *statsContext
-	HTTPClient   *http.Client
-	replicator   *base.Replicator
+	config           *ServerConfig
+	databases_       map[string]*db.DatabaseContext
+	lock             sync.RWMutex
+	statsContext     *statsContext
+	HTTPClient       *http.Client
+	replicator       *base.Replicator
+	pprofFileHandler *os.File
```
Comment (bbrks) [inline rest/server_context.go]: If we're going with this approach, I think we need a mutex around it. It's possible for 2 HTTP requests to simultaneously read/write.

**Review by bbrks — CHANGES_REQUESTED** [review]

File: rest/server_context.go (hunk 1 above)
Comment (adamcfraser) [reply rest/server_context.go]: Based on discussion w/ Sarath, we believe this is a non-test issue on Windows - SG will keep the file open for asynchronous requests run on Windows.

File: rest/api.go [hunk 2]
```diff
@@ -298,6 +302,7 @@ func (h *handler) handleProfiling() error {
 	if isCPUProfile {
 		base.Infof(base.KeyAll, "Starting CPU profile to %s ...", base.UD(params.File))
 		err = pprof.StartCPUProfile(f)
+		h.server.pprofFileHandler = f
```
Comment (adamcfraser) [inline rest/api.go]: If StartCPUProfile returns err!=nil (e.g. attempt to start a new profile while one is already running, with a different file name), we need to:
- return that error and not set pprofFileHandler
- remove the file from the filesystem

Even with that in place, we need additional locking for pprofFileHandler, as concurrent Start and Stop CPU profiling requests could race updating pprofFileHandler.  Should add a mutex for this handler, and write accessors to read/write the value to encapsulate the mutex handling.

File: rest/server_context.go (hunk 1 above)
Comment (sarathkumarsivan) [reply rest/server_context.go]: Fixed in latest revision

File: rest/api.go (hunk 2 above)
Comment (sarathkumarsivan) [reply rest/api.go]: Fixed in latest revision

File: rest/server_context.go [hunk 3]
```diff
@@ -46,6 +47,21 @@ type ServerContext struct {
 	statsContext *statsContext
 	HTTPClient   *http.Client
 	replicator   *base.Replicator
+	sync.Mutex            // Concurrent Start and Stop CPU profiling requests could race updating cpuPprofFile.
```
Comment (bbrks) [inline rest/server_context.go]: We want this lock to only affect access to the `cpuPprofFile` variable, so I don't think embedding the mutex makes sense here.

File: rest/api.go [hunk 4]
```diff
@@ -297,7 +301,13 @@ func (h *handler) handleProfiling() error {
...
 	if isCPUProfile {
 		base.Infof(base.KeyAll, "Starting CPU profile to %s ...", base.UD(params.File))
-		err = pprof.StartCPUProfile(f)
+		if err = pprof.StartCPUProfile(f); err != nil {
+			if fileError := os.Remove(params.File); fileError != nil {
+				base.Infof(base.KeyAll, "Error removing file: %s", base.UD(params.File))
+			}
+			return err
+		}
+		h.server.SetCpuPprofFile(nil)
```
Comment (bbrks) [inline rest/api.go]: Shouldn't this parameter be `f`?

File: rest/api.go [hunk 5]
```diff
@@ -285,6 +285,10 @@ func (h *handler) handleProfiling() error {
 		if isCPUProfile {
 			base.Infof(base.KeyAll, "... ending CPU profile")
 			pprof.StopCPUProfile()
+			if fileCloseError := h.server.GetCpuPprofFile().Close(); fileCloseError != nil {
+				base.Warnf("Error closing profile file: %v", fileCloseError)
+			}
+			h.server.SetCpuPprofFile(nil)
```
Comment (bbrks) [inline rest/api.go]: I still think there's a race here between start and stop, due to the lock being released between lines 288 and 291.

We'll need an operation to both close the existing file, and nil the file descriptor whilst holding the lock.

**Review by bbrks — CHANGES_REQUESTED** [review]

**Comment by bbrks** [comment]
Looks like we have a bunch of additional commits in here from some kind of merge.

Can we rebase onto master with a clean set of commits for this PR?

File: rest/server_context.go (hunk 3 above)
Comment (sarathkumarsivan) [reply rest/server_context.go]: Makes sense; A lock cpuPprofFileMutex has been added to protect cpuPprofFile from concurrent Start and Stop CPU profiling requests. Changes are available in latest revision.

File: rest/api.go (hunk 4 above)
Comment (sarathkumarsivan) [reply rest/api.go]: You are right. Change is made in latest revision.

File: rest/api.go (hunk 5 above)
Comment (sarathkumarsivan) [reply rest/api.go]: Closing the open file descriptor and pointing it to a nil reference can be implemented in CloseCpuPprofFile with acquiring and releasing the lock. Changes are available in latest revision. Warning the error whilst closing file descriptor is also included in this method.

**Review by sarathkumarsivan — COMMENTED** [review]
Revised the changes as per the latest review comments.

**Comment by bbrks** [comment]
I've force pushed a rebased version of this PR onto master to remove the commits I mentioned on Friday.

**Review by bbrks — APPROVED** [review]
The code looks good, but I've not had chance to test this on Windows due to the amount of setup required to get SG unit tests running there.

I think we should merge this as-is. If the tests work on Windows in the Jenkins builds, I can easily test the official Windows builds at that point.

**Pull request merged**